consensus/istanbul/validator/weighted.go

```diff
@@ -607,17 +656,62 @@ func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64) error
 	return nil
 }
 
+// updateValidators converts weighted validator slice to istanbul.Validators and sets them to the council.
+func (valSet *weightedCouncil) updateValidators(validators []*weightedValidator, demoted []*weightedValidator) {
+	var (
+		newValidators istanbul.Validators
+		newDemoted    istanbul.Validators
+	)
+
+	for _, val := range validators {
+		newValidators = append(newValidators, val)
+	}
+
+	for _, val := range demoted {
+		newDemoted = append(newDemoted, val)
+	}
+
+	valSet.validators = newValidators
+	valSet.demoted = newDemoted
+}
+
+// filterPoorValidators divided the given weightedValidators into two group filtered by the minimum amount of staking.
+func filterPoorValidators(weightedValidators []*weightedValidator, stakingAmounts []float64) ([]*weightedValidator, []float64, []*weightedValidator, []float64) {
+	var (
+		newWeightedValidators []*weightedValidator
+		newWeightedDemoted    []*weightedValidator
+		newValidatorsStaking  []float64
+		newDemotedStaking     []float64
+	)
+	amount := params.MinimumStakingAmount()
+	for idx, val := range stakingAmounts {
+		if uint64(val) >= amount.Uint64() {
+			newWeightedValidators = append(newWeightedValidators, weightedValidators[idx])
+			newValidatorsStaking = append(newValidatorsStaking, val)
+		} else {
+			newWeightedDemoted = append(newWeightedDemoted, weightedValidators[idx])
+			newDemotedStaking = append(newDemotedStaking, val)
+		}
+	}
+
+	if len(newWeightedValidators) <= 0 {
+		return newWeightedDemoted, newDemotedStaking, nil, nil
```

Comment:
How about returning an empty slice instead of `nil`? If use tried to call a method of the returned value, the node will be panicked.

Reply:
Updated. Thanks.